File: 2025-04-04/Binary_Tree_Zigzag_Level_Order_Traversal.java
package amazon;

import java.util.*;

public class Binary_Tree_Zigzag_Level_Order_Traversal {
    public static List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root == null) return new ArrayList<>();

        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        boolean isEven = false;

        while(!queue.isEmpty()) {
            int size = queue.size();

            List<Integer> cur = new ArrayList<>();
            for(int i=0;i<size;i++) {
                TreeNode node = queue.poll();
                cur.add(node.val);

                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }

            if(isEven)
                Collections.reverse(cur);

            isEven = !isEven;
            result.add(cur);
        }

        return result;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{1,2,3,4,null,null,5};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        System.out.println(zigzagLevelOrder(root));
    }
}

File: 2025-10-08/Binary_Tree_Zigzag_Level_Order_Traversal.py
from collections import deque
from typing import Optional, List

from TreeNode import TreeNode
from Tree import Tree


class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        result = []
        reversed = False
        q = deque()
        q.append(root)
        while q:
            l = len(q)
            curList = []
            for i in range(l):
                node = q.popleft()
                curList.append(node.val)

                if node.left:
                    q.append(node.left)

                if node.right:
                    q.append(node.right)

            if reversed:
                curList.reverse()

            result.append(curList)
            reversed = not reversed

        return result

sol = Solution()
tree = Tree()
node = tree.array_to_tree([3,9,20,None,None,15,7])
print(sol.zigzagLevelOrder(node))
